File: TheaterSuggestions/CSharp/SeatsSuggestions.Infra/Adapter/AuditoriumSeatingAdapter.cs
using System.Collections.Generic;
using System.Threading.Tasks;
using ExternalDependencies;
using SeatsSuggestions.Domain;
using SeatsSuggestions.Domain.Ports;

namespace SeatsSuggestions.Infra.Adapter
{
    /// <summary>
    ///     Adapt DTO instances coming from the external dependencies and Bounded Contexts (ReservationsProvider, AuditoriumLayoutRepository)
    ///     to instances of AuditoriumSeating (i.e. ).
    /// </summary>
    public class AuditoriumSeatingAdapter : IProvideUpToDateAuditoriumSeating
    {
        private readonly IProvideCurrentReservations _seatsReservationsProviderWebClient;
        private readonly IProvideAuditoriumLayouts _auditoriumLayoutProviderWebClient;

        public async Task<AuditoriumSeating> GetAuditoriumSeating(ShowId showId)
        {
            // Call the AuditoriumLayout Bounded Context in order to get an empty AuditoriumSeating
            var auditoriumDtoTask = _auditoriumLayoutProviderWebClient.GetAuditoriumSeatingFor(showId.Id);

            // Call the SeatReservation Bounded Context to get the list of already reserved seats
            var reservedSeatsDtoTask = _seatsReservationsProviderWebClient.GetReservedSeats(showId.Id);

            await Task.WhenAll(auditoriumDtoTask, reservedSeatsDtoTask);

            var auditoriumDto = auditoriumDtoTask.Result;
            var reservedSeatsDto = reservedSeatsDtoTask.Result;

            // Adapt all these information into a type belonging to our SeatSuggestion Bounded Context (External Domains/BCs INFRA => Our Domain)
            var auditoriumSeating = AdaptAuditoriumSeatingDto(auditoriumDto, reservedSeatsDto);

            return auditoriumSeating;
        }

        public AuditoriumSeatingAdapter(IProvideAuditoriumLayouts auditoriumLayoutProviderWebClient, IProvideCurrentReservations seatsReservationsProviderWebClient)
        {
            _auditoriumLayoutProviderWebClient = auditoriumLayoutProviderWebClient;
            _seatsReservationsProviderWebClient = seatsReservationsProviderWebClient;
        }


        /// <summary>
        /// Adapt 2 external BCs into 1 (our BC).
        /// </summary>
        /// <param name="auditoriumDto">The topology of an Auditorium coming from the AuditoriumLayout.API</param>
        /// <param name="reservedSeatsDto">The list of already reserved seats coming from the SeatReservations.API</param>
        /// <returns>An instance of <see cref="AuditoriumSeating"/> (i.e. The topology of the Auditorium with seats availability mapped)</returns>
        private static AuditoriumSeating AdaptAuditoriumSeatingDto(AuditoriumDto auditoriumDto, ReservedSeatsDto reservedSeatsDto)
        {
            var rows = new Dictionary<string, Row>();

            foreach (var (rowName, seatDtos) in auditoriumDto.Rows)
            {
                var seats = new List<Seat>();
                foreach (var seatDto in seatDtos)
                {
                    var number = ExtractSeatNumber(seatDto.Name);
                    var priceCategory = ConvertCategory(seatDto.Category);

                    var isReservationsSeat = reservedSeatsDto.ReservedSeats.Contains(seatDto.Name);
                    var seatAvailability = isReservationsSeat ? SeatAvailability.Reserved : SeatAvailability.Available;

                    seats.Add(new Seat(rowName, number, priceCategory, seatAvailability));
                }

                rows[rowName] = new Row(rowName, seats);
            }

            return new AuditoriumSeating(rows);
        }

        private static PricingCategory ConvertCategory(int dtoPricingCategory)
        {
            return (PricingCategory) dtoPricingCategory;
        }

        private static uint ExtractSeatNumber(string seatName)
        {
            return uint.Parse(seatName.Substring(1));
        }
    }
}

File: TheaterSuggestions/Java/SeatsSuggestionsInfra/src/main/java/com/baasie/SeatsSuggestionsInfra/AuditoriumSeatingAdapter.java
package com.baasie.SeatsSuggestionsInfra;

import com.baasie.SeatsSuggestionsDomain.*;

import com.baasie.ExternalDependencies.IProvideAuditoriumLayouts;
import com.baasie.ExternalDependencies.IProvideCurrentReservations;
import com.baasie.ExternalDependencies.auditoriumlayoutrepository.AuditoriumDto;
import com.baasie.ExternalDependencies.auditoriumlayoutrepository.SeatDto;
import com.baasie.ExternalDependencies.reservationsprovider.ReservedSeatsDto;
import com.baasie.SeatsSuggestionsDomain.port.IAdaptAuditoriumSeating;
import com.google.common.collect.ImmutableList;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class AuditoriumSeatingAdapter implements IAdaptAuditoriumSeating {

    private final IProvideCurrentReservations reservedSeatsRepository;
    private final IProvideAuditoriumLayouts auditoriumLayoutRepository;


    public AuditoriumSeatingAdapter(IProvideAuditoriumLayouts auditoriumLayoutRepository, IProvideCurrentReservations reservationsProvider) {
        this.auditoriumLayoutRepository = auditoriumLayoutRepository;
        this.reservedSeatsRepository = reservationsProvider;
    }

    private static PricingCategory convertCategory(int seatDtoCategory) {
        return PricingCategory.valueOf(seatDtoCategory);
    }

    private static int extractNumber(String name) {
        return Integer.parseUnsignedInt(name.substring(1));
    }

    public AuditoriumSeating getAuditoriumSeating(ShowId showId) {
        return adapt(auditoriumLayoutRepository.getAuditoriumSeatingFor(showId.ID()),
                reservedSeatsRepository.getReservedSeats(showId));

    }

    private AuditoriumSeating adapt(AuditoriumDto auditoriumDto, ReservedSeatsDto reservedSeatsDto) {

        Map<String, Row> rows = new HashMap<>();

        for (Map.Entry<String, ImmutableList<SeatDto>> rowDto : auditoriumDto.rows.entrySet()) {
            List<Seat> seats = new ArrayList<>();

            rowDto.getValue().forEach(seatDto -> {
                String rowName = rowDto.getKey();
                int number = extractNumber(seatDto.name);
                PricingCategory pricingCategory = convertCategory(seatDto.category);

                boolean isReserved = reservedSeatsDto.reservedSeats.contains(seatDto.name);

                seats.add(new Seat(rowName, number, pricingCategory, isReserved ? SeatAvailability.Reserved : SeatAvailability.Available));
            });

            rows.put(rowDto.getKey(), new Row(rowDto.getKey(), seats));
        }

        return new AuditoriumSeating(rows);
    }
}